localstack/config.py

```diff
@@ -584,18 +549,55 @@ def legacy_fallback(envar_name: str, default: T) -> T:
 @dataclass
 class HostAndPort:
     host: str
-    port: int
+    port: Optional[int]
 
     @classmethod
-    def parse(cls, input: str) -> "HostAndPort":
-        host, port_s = input.split(":")
-        return cls(host=host, port=int(port_s))
-
-    def __eq__(self, other: Any) -> bool:
-        if not isinstance(other, self.__class__):
-            return False
-
-        return self.host == other.host and self.port == other.port
+    def parse(
+        cls,
+        input: str,
+        default_host: Optional[str] = None,
+        default_port: int = constants.DEFAULT_PORT_EDGE,
+    ) -> "HostAndPort":
+        host, port = default_host, default_port
+        if ":" in input:
+            hostname, port_s = input.split(":", 1)
+            if hostname.strip():
+                host = hostname.strip()
+            try:
+                port = int(port_s)
+            except (ValueError, TypeError):
+                pass
+        else:
+            if input.strip():
+                host = input.strip()
+
+        # validation
+        if host is None:
+            raise ValueError("could not parse hostname")
+
+        if port < 0 or port >= 2**16:
+            raise ValueError("port out of range")
+
+        return cls(host=host, port=port)
+
+    def is_unprivileged(self) -> Optional[bool]:
+        if self.port is not None:
```

Comment:
Why can `self.port` not be `None`? `HostAndPort` is a dataclass, and someone may construct the type manually (e.g. `HostAndPort("0.0.0.0", port=None)`) so `port` could be `None`.

On second viewing of this PR (it's been a few months) I think `HostAndPort` should be made stricter - raising errors if the type is constructed incorrectly. This will mean a quicker exit for LS on errors, and as such more obvious feedback to the user if they supply invalid configuration.